app/controllers/tests_controller.rb

```diff
@@ -0,0 +1,51 @@
+class TestsController < ApplicationController
+ before_action :authenticate_teacher! ,only: [:create,:update,:destory,:edit]
+ before_action :authenticate_student! ,only: [:show]
+
+def index
+ @tests = Test.all
+end
+
+def new
+@test=Test.new
+end
+
+def create
+@test=Test.new(params_arg)
+if @test.save
+render:'show'
+else
+render:'new'
+end
+end
+
+def edit
+@test = Test.find(params[:id])
```

Comment:
`@test = Test.find(params[:id])`

is same in three methods edit, update and delete.

Set this as  a method and call it from before filter.

for example check  TeachersController -> set_teacher method